### ISSUE TITLE
Parameters inside function return type are not bound

### ISSUE THREAD

**Issue by muglug** [facebook/hhvm#8413]
Seen on

```
HipHop VM 3.30.1-0brew (rel)
hh-b2912d0d49dc93d6b9a5b0ff0605b4c5e11fc6ef-3.30.1 Tue Dec 18 13:35:22 2018
MacOS Mojave
```

### Examples 
```php
class A {}
class B {}

function gambleFn<T>(
  T $t1,
) : (function(T):T) {
  return function(T $t) use($t1) {
    if (rand(0, 1)) return $t1;
    return $t;
  };
}

gambleFn(new A())(new B());
```

### Actual result

'No errors!'

### Expected result

Some sort of invalid argument warning

**Comment by muglug** [comment]
uhh but that still doesn't work - `gambleFn<A>(new A())(new B())` produces no errors, even though it should

**Comment by manzyuk** [comment]
We are currently overhauling the type-checker using a new type inference algorithm based on solving subtyping constraints.  In this particular example, it would work something like this:
- Instantiate the type parameter `T` with a fresh type variable `v`.
- The call `gambleFn(new A())` gives rise to a constraint `A <: v` (the type of the argument must be a subtype of the declared type of the formal parameter), and the type of `gambleFn(new A())` is `(function(v): v)`.
- The call `gambleFn(new A())(new B())` gives rise to a constraint `B<: v`, and the type of `gambleFn(new A())(new B())` is `v`.

Therefore, we end up with two constraints `A <: v` and `B <: v`, and the smallest solution to them is `v = A | B` (union of `A` and `B`, currently represented as an unresolved type internally).  This makes sense: `gambleFn($x)($y)` returns either `$x` or `$y`, so its type should be the union of the types of `$x` and `$y`.

**Comment by muglug** [comment]
Hmm. That would make a lot of sense if the function was just

```php
function gamble<T>(T $t1, T $t2) : T {...}
```

But what if my intent was prevent someone calling `gambleFn` with any non-`A`? It seems odd to leak the constraint beyond the scope of the templated function.

It’s also been over a decade since I studied any type theory (I’m going to spend a couple of weeks with Types and Programming Languages to remedy that) so forgive me if I’m barking up the wrong tree.

**Comment by muglug** [comment]
> Your `gamble` is just an uncurried version of gambleFn.

Yeah, I just thought that might change how it was approached by the typechecker. Anyway, thanks again for the explanation. I see now that Flow uses the subtyping constraint system mentioned above.

**Comment by muglug** [comment]
I think this should still be an error, though?
 
```php
function bar<T>(
  (function((function():T)):T) $closure
) : void {}

bar(
  function(
    (function():int) $s
  ) : string {
    return "hello" . $s();
  }
);
```

**Comment by manzyuk** [comment]
This does produce an error on master.

**Comment by muglug** [comment]
Awesome!
